### PR TITLE
[WIP]: Crds integration

### PR DESCRIPTION
**Pull request by bhilbert4**
This PR integrates the CRDS package into Mirage.

When running `catalog_seed_generator`, `dark_prep`, or `observation_generator`, for any calibration reference file listed in the input yaml file, if its value is set to 'crds', Mirage will now query CRDS for the best reference files to use for a given exposure. If the reference file is not present locally, it will be downloaded, and then used.

`yaml_generator` has been updated to support this, via two new keywords. `reffile_defaults` can be set to 'crds' or 'crds_full_name'. 

If it is set to 'crds', then when run, yaml_generator will insert the string 'crds' into all of the jwst calibration reference file entries in the yaml files being produced. In this case, a given yaml file will always use the most up-to-date reference files regardless of when it is used by Mirage, since the CRDS query is done while creating the simulated data.

If `reffile_defaults` is set to 'crds_full_name', then the **yaml_generator** will query CRDS while creating the yaml files. In this case, the actual reference file names returned by the query will be placed into the yaml files. This leads to yaml files where it is easier to track exactly which reference files are used to create a given simulated exposure. It also means that no matter when that yaml file is used by Mirage, the output should be the same (modulo noise and CRs.). 

The `reffile_overrides` keyword has also been added to the `yaml_generator`. This can be given a nested dictionary specifying reference files to use for particular instrument configurations. In this way, a user can ignore the reference files returned by the CRDS query and use whatever they would like.

A few related minor changes: 

Users may now set the CRDS_PATH environment variable to the location on their local machine where they would like reference files downloaded to. (Note that this is required by the CRDS package). If this env variable is not set, then Mirage sets it to the default location of $HOME/crds_cache

CRDS tools have been added in the `reference_files/crds_tools.py` module.

Documentation on the `yaml_generator` was updated to describe these changes.

The `full_paths` function was pulled out of the seed generator, dark_prep, and observation_generator, and placed into utils.

A mapping of exposure types to instrument modes was added to constants.

### CODE REVIEW

**Comment by bhilbert4** [comment]
Closes #75

**Comment by bhilbert4** [comment]
I'm getting some pressure from the MO to finally get Mirage publicly released. I'm working on generating updated gridded PSF libraries at the moment, and these will go onto Box and and my downloader script will be modified to point to them. The other thing I'd like to do is get this PR merged so that we don't need to maintain an up-to-date set of pipeline reference files. Does anyone have a few minutes to look over these changes/try them out?

**Comment by mperrin** [comment]
@bhilbert4 I'm definitely not the right person to look over the CRDS stuff, but perhaps either I or Marcio or Shannon should look over your updated PSF libraries and help cross check those? 
Are you using the latest development version of webbpsf for that?   (I merged a big PR on Aug 15 that improves the field dependence model in a few ways... )

**Comment by Johannes-Sahlmann** [comment]
I am trying it out and find a problem with the PSF library: 
I believe that [with-path]/fgs/gridded_psf_library/psf_wings/ has too many files. Some are names `niriss_fgs1_[...]` which is probably incorrect?

**Comment by bhilbert4** [comment]
New NIRISS libraries are being produced as we speak. I'll remove the old files shortly.

**Comment by Johannes-Sahlmann** [comment]
Thanks @bhilbert4 I am passing the PSF step now. 
The next error is related to CRDS:
```
CRDS - ERROR -  (FATAL) CRDS server connection and cache load FAILED.  Cannot continue.
CRDS - ERROR -  See https://hst-crds.stsci.edu/docs/cmdline_bestrefs/ or https://jwst-crds.stsci.edu/docs/cmdline_bestrefs/
CRDS - ERROR -  for more information on configuring CRDS,  particularly CRDS_PATH and CRDS_SERVER_URL. : [Errno 2] No such file or directory: '/Users/jsahlmann/crds_cache/config/jwst/server_config'
```
Do I have to create that `crds_cache/config/jwst/server_config` file?

**Comment by bhilbert4** [comment]
No, you shouldn't have to create that directory. This error has to do with the `crds` package and imports. If you import `crds` prior to setting the environment variables, then this error pops up. The difficult part is that it means you can't import anything else that may import `crds` either, such as the `jwst` package. 

I fought with this for a while when putting everything into mirage to be sure that `crds` is not imported until after Mirage sets the environment variables. I may have missed something though. Did you import the pipeline (or some other package that imports crds) prior to running mirage?

**Comment by Johannes-Sahlmann** [comment]
I see. Yes, I am importing modules from the `jwst` package prior to running mirage. I would have to refactor my code to separate these things, which I would like to avoid.
What would be the easiest way to trigger the setting of the Mirage environment variables within an external script?

**Comment by bhilbert4** [comment]
Try explicitly setting CRDS_PATH and CRDS_SERVER_URL prior to your jwst imports. I wrote `crds_tools.py` such that if the env variables are already set, Mirage doesn't change them. So that should get around the import problem. I'm not sure there's another way around this at the moment. The `crds` developers know that it's a problem, but I'm not sure if there's a timeline to fix it.

**Comment by Johannes-Sahlmann** [comment]
Thanks. Prepending
```
os.environ["CRDS_PATH"] = '{}/crds_cache'.format(os.environ.get('HOME'))
os.environ["CRDS_SERVER_URL"] = "https://jwst-crds.stsci.edu"
```
to my `jwst` imports fixed that issue.

Now I am getting `crds.core.exceptions.CrdsLookupError: Error determining best reference for 'mask'  =    parameter='META.INSTRUMENT.DETECTOR' value='FGS1' is not in ['GUIDER1', 'GUIDER2']`
which is a different problem with probably has a simple fix. I will investigate this tomorrow.

**Comment by Johannes-Sahlmann** [comment]
I ran this branch on a test program with NIRISS and FGS imaging and it seemed to work fine after the minor fix for which I put in a PR against your fork. I did not yet run the products through the pipeline but I don't expect problems with that. 
I will approve this PR after you have merged my mini PR.

File: docs/yaml_generator.rst [hunk 1]
```diff
@@ -73,6 +73,86 @@ There are currently a number of parameters which can only be set using the ``par
...
+
+.. tip::
+
+    In order to specify a location on your machine to store the downloaded reference files, you must have the CRDS_PATH environment variable set to that location. If this environment variable is not set, Mirage will default to use $HOME/crds_cache/
+
+It is also possible to specify that Mirage use reference files other than those downloaded from CRDS. In order to do this, you must supply a dictionary of filenames. Due to the varying number of selection criteria needed to uniquely identify the reference file that matches up with a particular exposure, this dictionary is composed of multiple levels of nested dictionaries. Not all possibilities are required. You may specify reference files only for the particular observing modes you are interested in. Any modes in your APT file that are not contained within the dictionary will revert to using the reference files identified by CRDS. Below is a dictionary showing all nesting required for all reference files. Note that the dictionary structure is instument dependent since a reference file type for different instruments does not necessarily have the same selection criteria.
```
[the remark this answers is not in the document]
Comment (bhilbert4) [reply docs/yaml_generator.rst]: Corrected

File: docs/yaml_generator.rst [hunk 2]
```diff
@@ -73,6 +73,86 @@ There are currently a number of parameters which can only be set using the ``par
...
+
+.. tip::
+    If you choose to provide your own reference files, it is best to use these same reference files when running the JWST calibration pipeline on the simualted data files produced by Mirage. Not doing this can lead to systematic errors in your calibrated data.
```
[the remark this answers is not in the document]
Comment (bhilbert4) [reply docs/yaml_generator.rst]: Corrected

**Comment by bhilbert4** [comment]
@mperrin I grabbed the latest versions of webbpsf and poppy last week, after I saw that the NIRCam Zernike and focus updates were merged. I'm using webbpsf version 0.8.0 and poppy version 0.9.0dev1387.dev1387 to create the gridded psf libraries. 

I'm going to merge this CRDS PR, and create a separate issue to discuss gridded libraries, but yes, I'd love it if you or Shannon could have a look at the libraries and make sure they make sense.

**Pull request merged**